inginious/frontend/pages/lti.py

```diff
@@ -52,78 +54,175 @@ class LTIBindPage(INGIniousAuthPage):
+        try:
+            course = self.course_factory.get_course(courseid)
+        except exceptions.CourseNotFoundException as ex:
+            raise NotFound(description=_(str(ex)))
+
+        flask_request = FlaskRequest()
+        target_link_uri = flask_request.get_param('target_link_uri')
+        if not target_link_uri:
+            raise Exception('Missing "target_link_uri" param')
+        taskid = target_link_uri.split('/')[-1]
+
+        launch_data_storage = MongoLTILaunchDataStorage(self.database, courseid, taskid)
+        oidc_login = FlaskOIDCLogin(flask_request, course.lti_tool(), launch_data_storage=launch_data_storage)
+        return oidc_login.enable_check_cookies().redirect(target_link_uri)
+
+    def GET(self, courseid):
+        return self._handle_oidc_login_request(courseid)
+
+    def POST(self, courseid):
+        return self._handle_oidc_login_request(courseid)
+
+
+class LTILaunchPage(INGIniousPage):
+    endpoint = 'ltilaunchpage'
+
+    def _handle_message_launch(self, courseid, taskid):
+        """ Decrypt and process the LTI Launch message. """
+        try:
+            course = self.course_factory.get_course(courseid)
+        except exceptions.CourseNotFoundException as ex:
+            raise NotFound(description=_(str(ex)))
+
+        tool_conf = course.lti_tool()
+        launch_data_storage = MongoLTILaunchDataStorage(self.database, courseid, taskid)
+        flask_request = FlaskRequest()
+        message_launch = FlaskMessageLaunch(flask_request, tool_conf, launch_data_storage=launch_data_storage)
+
+        launch_id = message_launch.get_launch_id()  # TODO(mp): With a good use of the cache, this could be used as a non-session id
+        launch_data = message_launch.get_launch_data()
+
+        user_id = launch_data['sub']
+        roles = launch_data['https://purl.imsglobal.org/spec/lti/claim/roles']
+        realname = self._find_realname(launch_data)
+        email = launch_data.get('email', '')
+        platform_instance_id = '/'.join([launch_data['iss'], message_launch.get_client_id(), launch_data['https://purl.imsglobal.org/spec/lti/claim/deployment_id']])
+        tool = launch_data.get('https://purl.imsglobal.org/spec/lti/claim/tool_platform', {})
+        tool_name = tool.get('name', 'N/A')
+        tool_desc = tool.get('description', 'N/A')
+        tool_url = tool.get('url', 'N/A')
+        context = launch_data['https://purl.imsglobal.org/spec/lti/claim/context']
+        context_title = context.get('context_title', 'N/A')
+        context_label = context.get('context_label', 'N/A')
+
+        can_report_grades = message_launch.has_ags() and tool_conf.get_iss_config(iss=message_launch.get_iss(),
+                                                                                  client_id=message_launch.get_client_id()).get('auth_token_url')
+
+        session_id = hashlib.sha256(launch_id.encode('utf-8')).digest().hex()  # TODO(mp): Make this more secure
```

Comment:
This depends on what the launch id is based on. However, as it is probably shared externally, it would indeed be more secure to add some private hint in the session id generation to avoid session hijacking. Maybe based on the flask session secret parameter.